### PR TITLE
rgw multisite: teach 'bucket sync checkpoint' about log generations

### PR DESCRIPTION
**Pull request by cbodley**
`radosgw-admin bucket sync checkpoint` is used by multisite tests to wait for a bucket sync to catch up with a source zone. sync on a resharded bucket is not caught up until the sync status' generation matches the source zone's latest_gen

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/rgw/services/svc_bi_rados.cc [hunk 1]
```diff
@@ -163,23 +163,23 @@ static void get_bucket_index_objects(const string& bucket_oid_base,
 }
 
 static void get_bucket_instance_ids(const RGWBucketInfo& bucket_info,
-                                    int shard_id,
+                                    int num_shards, int shard_id,
```
Comment (yuvalif) [inline src/rgw/services/svc_bi_rados.cc:166]: why do you pass "num_shards" if this could be fetched from "bucket_info"?

File: src/rgw/rgw_data_sync.cc [hunk 2]
```diff
@@ -5520,17 +5520,11 @@ int rgw_read_remote_bilog_info(const DoutPrefixProvider *dpp,
...
     { nullptr, nullptr }
   };
-  rgw_bucket_index_marker_info result;
-  int r = conn->get_json_resource(dpp, "/admin/log/", params, y, result);
+  int r = conn->get_json_resource(dpp, "/admin/log/", params, y, info);
   if (r < 0) {
     ldpp_dout(dpp, -1) << "failed to fetch remote log markers: " << cpp_strerror(r) << dendl;
     return r;
   }
-  r = markers.from_string(result.max_marker, -1);
```
Comment (yuvalif) [inline src/rgw/rgw_data_sync.cc:5529]: how is this change related to the generation fix?
you call "from_string" on the result of the function in 2 places. isn't it batter to keep it inside "rgw_read_remote_bilog_info()" ?

File: src/rgw/rgw_sync_checkpoint.cc [hunk 3]
```diff
@@ -115,6 +112,29 @@ int bucket_source_sync_checkpoint(const DoutPrefixProvider* dpp,
     }
   }
 
+  // wait for latest_gen
+  while (full_status.incremental_gen < latest_gen) {
+    auto delay_until = ceph::coarse_mono_clock::now() + retry_delay;
```
Comment (yuvalif) [inline src/rgw/rgw_sync_checkpoint.cc:117]: nit: const

File: src/rgw/services/svc_bi_rados.cc (hunk 1 above)
Comment (cbodley) [reply src/rgw/services/svc_bi_rados.cc:166]: this function was hard-coded to use num_shards from the `current_index` layout, but this function is called by `open_bucket_index()` which takes an arbitrary `rgw::bucket_index_layout_generation` so we need to use the shards from that index layout instead of `current_index`

this is relevent now because we can request `RGWOp_BILog_Info` for any log generation; we convert that to an index layout with `log_to_index_layout()` and pass that to `RGWRados::get_bucket_stats()` which calls into `open_bucket_index()`

without this change, we hit an assert in `RGWRados::get_bucket_stats()` because `open_bucket_index()` returns a different number of `bucket_objs` (based on the given index's num_shards) vs `bucket_instance_ids` (based on current_index num_shards)

File: src/rgw/rgw_data_sync.cc (hunk 2 above)
Comment (cbodley) [reply src/rgw/rgw_data_sync.cc:5529]: the callers of `rgw_read_remote_bilog_info()` now need more than just `info.max_marker`, so i changed it to return `rgw_bucket_index_marker_info` directly instead of `BucketIndexShardsManager`. but i can change it to return both so the callers don't have to duplicate the marker parsing

**Comment by cbodley** [comment]
thanks @yuvalif, updated

**Review by yuvalif — APPROVED** [review]

File: src/rgw/rgw_sync_checkpoint.cc [hunk 4]
```diff
@@ -115,6 +112,36 @@ int bucket_source_sync_checkpoint(const DoutPrefixProvider* dpp,
...
+    if (r < 0 && r != -ENOENT) { // retry on ENOENT
+      return r;
+    }
+  }
+
+  if (full_status.incremental_gen > latest_gen) {
+    ldpp_dout(dpp, 1) << "bucket sync caught up with source:\n"
+        << "        local gen: " << full_status.incremental_gen << '\n'
+        << "       remote gen: " << latest_gen << dendl;
+    return 0;
```
Comment (cbodley) [inline src/rgw/rgw_sync_checkpoint.cc:137]: added this last bit in case we overshoot the remote's starting generation, so we don't go on and try to compare markers from different generations

File: src/rgw/services/svc_bi_rados.cc (hunk 1 above)
Comment (smanjara) [reply src/rgw/services/svc_bi_rados.cc:166]: thanks. I should have changed this as part of the bilog pr.

**Review by smanjara — APPROVED** [review]
